src/ceph-daemon

```diff
@@ -0,0 +1,1311 @@
+        pc = CephContainer(
+            image=args.image,
+            entrypoint='/usr/sbin/ceph-volume',
+            args=[
+                'lvm', 'activate',
+                daemon_id, args.osd_fsid,
+                '--no-systemd'
+            ],
+            podman_args=['--privileged'],
+            volume_mounts=get_container_mounts(fsid, daemon_type, daemon_id),
+            cname='ceph-%s-activate-%s.%s' % (fsid, daemon_type, daemon_id),
+        )
+        pc.run()
+
+    deploy_daemon_units(fsid, daemon_type, daemon_id, c)
+
+def deploy_daemon_units(fsid, daemon_type, daemon_id, c,
+                        enable=True, start=True):
+    # cmd
+    data_dir = get_data_dir(fsid, daemon_type, daemon_id)
+    with open(data_dir + '/cmd', 'w') as f:
+        f.write('#!/bin/sh\n' + ' '.join(c.run_cmd()) + '\n')
+        os.fchmod(f.fileno(), 0o700)
+
+    # systemd
+    install_base_units(fsid)
+    unit = get_unit_file(fsid)
+    unit_file = 'ceph-%s@.service' % (fsid)
+    with open(args.unit_dir + '/' + unit_file + '.new', 'w') as f:
+        f.write(unit)
+        os.rename(args.unit_dir + '/' + unit_file + '.new',
+                  args.unit_dir + '/' + unit_file)
+    subprocess.check_output(['systemctl', 'daemon-reload'])
+
+    unit_name = get_unit_name(fsid, daemon_type, daemon_id)
+    if enable:
+        subprocess.check_output(['systemctl', 'enable', unit_name])
+    if start:
+        subprocess.check_output(['systemctl', 'start', unit_name])
+
+def install_base_units(fsid):
+    """
+    Set up ceph.target and ceph-$fsid.target units.
+    """
+    existed = os.path.exists(args.unit_dir + '/ceph.target')
+    with open(args.unit_dir + '/ceph.target.new', 'w') as f:
+        f.write('[Unit]\n'
+                'Description=all ceph service\n'
+                '[Install]\n'
+                'WantedBy=multi-user.target\n')
+        os.rename(args.unit_dir + '/ceph.target.new',
+                  args.unit_dir + '/ceph.target')
+    if not existed:
+        subprocess.check_output(['systemctl', 'enable', 'ceph.target'])
+        subprocess.check_output(['systemctl', 'start', 'ceph.target'])
+
+    existed = os.path.exists(args.unit_dir + '/ceph-%s.target' % fsid)
+    with open(args.unit_dir + '/ceph-%s.target.new' % fsid, 'w') as f:
+        f.write('[Unit]\n'
+                'Description=ceph cluster {fsid}\n'
+                'PartOf=ceph.target\n'
+                'Before=ceph.target\n'
+                '[Install]\n'
+                'WantedBy=multi-user.target ceph.target\n'.format(
+                    fsid=fsid)
+        )
+        os.rename(args.unit_dir + '/ceph-%s.target.new' % fsid,
+                  args.unit_dir + '/ceph-%s.target' % fsid)
+    if not existed:
+        subprocess.check_output(['systemctl', 'enable', 'ceph-%s.target' % fsid])
+        subprocess.check_output(['systemctl', 'start', 'ceph-%s.target' % fsid])
+
+def get_unit_file(fsid):
+    u = """[Unit]
```

Comment:
Should these not be written to /usr/lib/systemd/... so the admin can use local overwrites via /etc/systemd/...?

Also, should these be embedded into the python code rather than packaged as template files?

Reply:
> Should these not be written to /usr/lib/systemd/... so the admin can use local overwrites via /etc/systemd/...?

I don't think so -- what is produced here effectively *is* a local overwrite already.

Reply:
exactly

Reply:
Also, this is by design a single standalone script so that you can (1) wget + run it, and (2) so that the ssh orchestrator can run it on remote hosts by passing it to /usr/bin/python3's stdin